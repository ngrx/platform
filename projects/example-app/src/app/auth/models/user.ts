export interface Credentials {
  username: string;
  password: string;
}

export interface User {
  name: string;
}
