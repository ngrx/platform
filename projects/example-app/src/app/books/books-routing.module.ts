import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import {
  CollectionPageComponent,
  FindBookPageComponent,
  ViewBookPageComponent,
} from '@example-app/books/containers';
import { BookExistsGuard } from '@example-app/books/guards';

export const routes: Routes = [
  {
    path: 'find',
    component: FindBookPageComponent,
    data: { title: 'Find book' },
  },
  {
    path: ':id',
    component: ViewBookPageComponent,
    canActivate: [BookExistsGuard],
    data: { title: 'Book details' },
  },
  {
    path: '',
    component: CollectionPageComponent,
    data: { title: 'Collection' },
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class BooksRoutingModule {}
