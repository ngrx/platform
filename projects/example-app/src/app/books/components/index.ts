export * from './book-authors.component';
export * from './book-detail.component';
export * from './book-preview.component';
export * from './book-preview-list.component';
export * from './book-search.component';
