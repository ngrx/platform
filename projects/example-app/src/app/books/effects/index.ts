export * from './book.effects';
export * from './collection.effects';
