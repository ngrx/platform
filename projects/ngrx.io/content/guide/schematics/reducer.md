# Reducer

---

## Overview

Generates a reducer file that contains a state interface,
an initial state object for the reducer, and a reducer function.

## Command

```sh
ng generate reducer ReducerName [options]
```

##### OR

```sh
ng generate r ReducerName [options]
```

### Options

Provide the project name where the reducer files will be created.

- `--project`
  - Alias: `-p`
  - Type: `string`

Nest the reducer file within a folder based on the reducer `name`.

- `--flat`
  - Type: `boolean`
  - Default: `true`

Group the reducer file within a `reducers` folder.

- `--group`
  - Alias: `-g`
  - Type: `boolean`
  - Default: `false`

Provide the path to a file containing an `Angular Module` and the entity reducer will be added to its `imports` array using `StoreModule.forFeature`.

- `--module`
  - Alias: `-m`
  - Type: `string`

Provide the path to a `reducers` file containing a state interface and an object map of action reducers. The generated reducer interface will be imported and added to the first defined interface within the file. The reducer will be imported and added to the first defined object with an `ActionReducerMap` type.

- `--reducers`
  - Alias: `-r`
  - Type: `string`

Specifies if api success and failure actions should be added to the reducer.

- `--api`
  - Alias: `-a`
  - Type: `boolean`
  - Default: `false`

Generate a spec file alongside the reducer file.

- `--skip-tests`
  - Type: `boolean`
  - Default: `false`

## Examples

Generate a `User` reducer file add it to a defined map of reducers generated from a [feature state](guide/schematics/store#examples).

```sh
ng generate reducer User --reducers reducers/index.ts
```

Generate a `User` reducer file within a nested folder based on the reducer name.

```sh
ng generate reducer User --flat false
```

Generate a `User` reducer and register it within the `Angular Module` in `app.module.ts`.

```sh
ng generate reducer User --module app.module.ts
```

Generate a `User` reducer file within a nested `reducers` folder.

```sh
ng generate reducer User --group
```
