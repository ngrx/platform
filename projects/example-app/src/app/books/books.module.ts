import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';

import { ComponentsModule } from '@example-app/books/components';
import { BookEffects } from '@example-app/books/effects/book.effects';
import { CollectionEffects } from '@example-app/books/effects/collection.effects';

import { FindBookPageComponent } from '@example-app/books/containers/find-book-page.component';
import { ViewBookPageComponent } from '@example-app/books/containers/view-book-page.component';
import { SelectedBookPageComponent } from '@example-app/books/containers/selected-book-page.component';
import { CollectionPageComponent } from '@example-app/books/containers/collection-page.component';
import { MaterialModule } from '@example-app/material';

import { reducers } from '@example-app/books/reducers';
import { BooksRoutingModule } from '@example-app/books/books-routing.module';

@NgModule({
  imports: [
    CommonModule,
    MaterialModule,
    ComponentsModule,
    BooksRoutingModule,

    /**
     * StoreModule.forFeature is used for composing state
     * from feature modules. These modules can be loaded
     * eagerly or lazily and will be dynamically added to
     * the existing state.
     */
    StoreModule.forFeature('books', reducers),

    /**
     * Effects.forFeature is used to register effects
     * from feature modules. Effects can be loaded
     * eagerly or lazily and will be started immediately.
     *
     * All Effects will only be instantiated once regardless of
     * whether they are registered once or multiple times.
     */
    EffectsModule.forFeature([BookEffects, CollectionEffects]),
  ],
  declarations: [
    FindBookPageComponent,
    ViewBookPageComponent,
    SelectedBookPageComponent,
    CollectionPageComponent,
  ],
})
export class BooksModule {}
