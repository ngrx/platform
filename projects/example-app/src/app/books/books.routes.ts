import { Routes } from '@angular/router';

import {
  CollectionPageComponent,
  FindBookPageComponent,
  ViewBookPageComponent,
} from '@example-app/books/containers';
import { bookExistsGuard } from '@example-app/books/guards';
import { provideState } from '@ngrx/store';
import * as fromBooks from '@example-app/books/reducers';
import { provideEffects } from '@ngrx/effects';
import { BookEffects, CollectionEffects } from './effects';

export const BOOKS_ROUTES: Routes = [
  {
    path: '',
    providers: [
      /**
       * provideState() is used for composing state
       * from feature modules. These modules can be loaded
       * eagerly or lazily and will be dynamically added to
       * the existing state.
       */
      provideState(fromBooks.booksFeatureKey, fromBooks.reducers),
      /**
       * provideEffects() is used to register effects
       * from feature modules. Effects can be loaded
       * eagerly or lazily and will be started immediately.
       *
       * All Effects will only be instantiated once regardless of
       * whether they are registered once or multiple times.
       */
      provideEffects(BookEffects, CollectionEffects),
    ],
    children: [
      {
        path: 'find',
        component: FindBookPageComponent,
        data: { title: 'Find book' },
      },
      {
        path: ':id',
        component: ViewBookPageComponent,
        canActivate: [bookExistsGuard],
        data: { title: 'Book details' },
      },
      {
        path: '',
        component: CollectionPageComponent,
        data: { title: 'Collection' },
      },
    ],
  },
];
