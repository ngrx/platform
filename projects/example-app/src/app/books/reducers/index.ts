import { Book } from '@example-app/books/models';
import {
  createSelector,
  createFeatureSelector,
  combineReducers,
  Action,
} from '@ngrx/store';
import * as fromSearch from '@example-app/books/reducers/search.reducer';
import * as fromBooks from '@example-app/books/reducers/books.reducer';
import * as fromCollection from '@example-app/books/reducers/collection.reducer';
import * as fromRoot from '@example-app/reducers';

export const booksFeatureKey = 'books';

export interface BooksState {
  [fromSearch.searchFeatureKey]: fromSearch.State;
  [fromBooks.booksFeatureKey]: fromBooks.State;
  [fromCollection.collectionFeatureKey]: fromCollection.State;
}

export interface State extends fromRoot.State {
  [booksFeatureKey]: BooksState;
}

/** Provide reducer in AoT-compilation happy way */
export function reducers(state: BooksState | undefined, action: Action) {
  return combineReducers({
    [fromSearch.searchFeatureKey]: fromSearch.reducer,
    [fromBooks.booksFeatureKey]: fromBooks.reducer,
    [fromCollection.collectionFeatureKey]: fromCollection.reducer,
  })(state, action);
}

/**
 * A selector function is a map function factory. We pass it parameters and it
 * returns a function that maps from the larger state tree into a smaller
 * piece of state. This selector simply selects the `books` state.
 *
 * Selectors are used with the `select` operator.
 *
 * ```ts
 * class MyComponent {
 *   constructor(state$: Observable<State>) {
 *     this.booksState$ = state$.pipe(select(getBooksState));
 *   }
 * }
 * ```
 */

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const selectBooksState = createFeatureSelector<BooksState>(
  booksFeatureKey
);

/**
 * Every reducer module exports selector functions, however child reducers
 * have no knowledge of the overall state tree. To make them usable, we
 * need to make new selectors that wrap them.
 *
 * The createSelector function creates very efficient selectors that are memoized and
 * only recompute when arguments change. The created selectors can also be composed
 * together to select different pieces of state.
 */
export const selectBookEntitiesState = createSelector(
  selectBooksState,
  (state) => state.books
);

export const selectSelectedBookId = createSelector(
  selectBookEntitiesState,
  fromBooks.selectId
);

/**
 * Adapters created with @ngrx/entity generate
 * commonly used selector functions including
 * getting all ids in the record set, a dictionary
 * of the records by id, an array of records and
 * the total number of records. This reduces boilerplate
 * in selecting records from the entity state.
 */
export const {
  selectIds: selectBookIds,
  selectEntities: selectBookEntities,
  selectAll: selectAllBooks,
  selectTotal: selectTotalBooks,
} = fromBooks.adapter.getSelectors(selectBookEntitiesState);

export const selectSelectedBook = createSelector(
  selectBookEntities,
  selectSelectedBookId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);

/**
 * Just like with the books selectors, we also have to compose the search
 * reducer's and collection reducer's selectors.
 */
export const selectSearchState = createSelector(
  selectBooksState,
  (state) => state.search
);

export const selectSearchBookIds = createSelector(
  selectSearchState,
  fromSearch.getIds
);
export const selectSearchQuery = createSelector(
  selectSearchState,
  fromSearch.getQuery
);
export const selectSearchLoading = createSelector(
  selectSearchState,
  fromSearch.getLoading
);
export const selectSearchError = createSelector(
  selectSearchState,
  fromSearch.getError
);

/**
 * Some selector functions create joins across parts of state. This selector
 * composes the search result IDs to return an array of books in the store.
 */
export const selectSearchResults = createSelector(
  selectBookEntities,
  selectSearchBookIds,
  (books, searchIds) => {
    return searchIds
      .map((id) => books[id])
      .filter((book): book is Book => book != null);
  }
);

export const selectCollectionState = createSelector(
  selectBooksState,
  (state) => state.collection
);

export const selectCollectionLoaded = createSelector(
  selectCollectionState,
  fromCollection.getLoaded
);
export const getCollectionLoading = createSelector(
  selectCollectionState,
  fromCollection.getLoading
);
export const selectCollectionBookIds = createSelector(
  selectCollectionState,
  fromCollection.getIds
);

export const selectBookCollection = createSelector(
  selectBookEntities,
  selectCollectionBookIds,
  (entities, ids) => {
    return ids
      .map((id) => entities[id])
      .filter((book): book is Book => book != null);
  }
);

export const isSelectedBookInCollection = createSelector(
  selectCollectionBookIds,
  selectSelectedBookId,
  (ids, selected) => {
    return !!selected && ids.indexOf(selected) > -1;
  }
);
