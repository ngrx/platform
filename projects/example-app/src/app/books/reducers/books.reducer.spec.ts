import { reducer } from '@example-app/books/reducers/books.reducer';
import * as fromBooks from '@example-app/books/reducers/books.reducer';
import {
  BooksApiActions,
  BookActions,
  ViewBookPageActions,
  CollectionApiActions,
} from '@example-app/books/actions';
import { Book, generateMockBook } from '@example-app/books/models';

describe('BooksReducer', () => {
  const book1 = generateMockBook();
  const book2 = { ...book1, id: '222' };
  const book3 = { ...book1, id: '333' };
  const initialState: fromBooks.State = {
    ids: [book1.id, book2.id],
    entities: {
      [book1.id]: book1,
      [book2.id]: book2,
    },
    selectedBookId: null,
  };

  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SEARCH_COMPLETE & LOAD_SUCCESS', () => {
    type BooksActions =
      | typeof BooksApiActions.searchSuccess
      | typeof CollectionApiActions.loadBooksSuccess;
    function noExistingBooks(
      action: BooksActions,
      booksInitialState: any,
      books: Book[]
    ) {
      const createAction = action({ books });

      const result = reducer(booksInitialState, createAction);

      expect(result).toMatchSnapshot();
    }

    function existingBooks(
      action: BooksActions,
      booksInitialState: any,
      books: Book[]
    ) {
      // should not replace existing books
      const differentBook2 = { ...books[0], foo: 'bar' };
      const createAction = action({ books: [books[1], differentBook2] });

      const expectedResult = {
        ids: [...booksInitialState.ids, books[1].id],
        entities: {
          ...booksInitialState.entities,
          [books[1].id]: books[1],
        },
        selectedBookId: null,
      };

      const result = reducer(booksInitialState, createAction);

      expect(result).toMatchSnapshot();
    }

    it('should add all books in the payload when none exist', () => {
      noExistingBooks(BooksApiActions.searchSuccess, initialState, [
        book1,
        book2,
      ]);

      noExistingBooks(CollectionApiActions.loadBooksSuccess, initialState, [
        book1,
        book2,
      ]);
    });

    it('should add only books when books already exist', () => {
      existingBooks(BooksApiActions.searchSuccess, initialState, [
        book2,
        book3,
      ]);

      existingBooks(CollectionApiActions.loadBooksSuccess, initialState, [
        book2,
        book3,
      ]);
    });
  });

  describe('LOAD', () => {
    const expectedResult = {
      ids: [book1.id],
      entities: {
        [book1.id]: book1,
      },
      selectedBookId: null,
    };

    it('should add a single book, if the book does not exist', () => {
      const action = BookActions.loadBook({ book: book1 });

      const result = reducer(fromBooks.initialState, action);

      expect(result).toMatchSnapshot();
    });

    it('should return the existing state if the book exists', () => {
      const action = BookActions.loadBook({ book: book1 });

      const result = reducer(expectedResult, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SELECT', () => {
    it('should set the selected book id on the state', () => {
      const action = ViewBookPageActions.selectBook({ id: book1.id });

      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('selectId', () => {
      it('should return the selected id', () => {
        const result = fromBooks.selectId({
          ...initialState,
          selectedBookId: book1.id,
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
