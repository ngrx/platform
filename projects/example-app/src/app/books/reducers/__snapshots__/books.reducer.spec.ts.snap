// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksReducer LOAD should add a single book, if the book does not exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedBookId": null,
}
`;

exports[`BooksReducer LOAD should return the existing state if the book exists 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedBookId": null,
}
`;

exports[`BooksReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all books in the payload when none exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedBookId": null,
}
`;

exports[`BooksReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all books in the payload when none exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedBookId": null,
}
`;

exports[`BooksReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only books when books already exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "333": Object {
      "id": "333",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedBookId": null,
}
`;

exports[`BooksReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only books when books already exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "333": Object {
      "id": "333",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedBookId": null,
}
`;

exports[`BooksReducer SELECT should set the selected book id on the state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedBookId": "1",
}
`;

exports[`BooksReducer Selectors selectId should return the selected id 1`] = `"1"`;

exports[`BooksReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "selectedBookId": null,
}
`;
